### ISSUE TITLE
Goal: More love, less hate

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#136]
I found the blog post [My love-hate relationship with LuaJIT](https://blog.separateconcerns.com/2015-06-18-love-hate-luajit.html) thought provoking. The title is hyperbolic but I can relate to the reflections. He nicely expresses both positive and negative sentiments about his experience as a LuaJIT user. I would like to reflect on how RaptorJIT can maximize the positive side and minimize the negative.

The positive side:

> Those are the features that make LuaJIT so attractive: its blazing speed and the FFI. They are why I could write Lua professionally for several years, and I could never thank Mike Pall enough for this. Moreover, LuaJIT has been playing a huge role in the growing popularity of Lua those last few years.

The negative side:

> Beyond the purist ideal of understanding the whole stack, this has practical implications. When you find a bug in LuaJIT, understanding it and fixing it is terribly complicated. The best I can do is usually to try to produce a small test case that reproduces the bug (even that is not always easy) and hope Mike Pall finds a fix. With PUC Lua I could probably fix it myself - but of course PUC Lua is so simple that I have never found a bug in it!

The social context:

> Another issue is that PUC Lua and LuaJIT are diverging. LuaJIT implements Lua 5.1. Lua 5.2 code can be made to work as long as it does not use _ENV, but code that leverages the new features in Lua 5.3 is not supported at all (although modules like lua-compat-5.3 can probably help). On the other hand, more and more modules require the FFI, making them incompatible with PUC Lua (of course I am a culprit of that myself). And to top it all, the interpreters work so differently that efficient code in LuaJIT is not necessarily good in PUC Lua, and vice versa.

### RaptorJIT

Here is a braindump on how I see this applying to RaptorJIT.

On the positive side, it is important to remember that LuaJIT is an absolutely awesome project, and that Mike Pall did the world a huge service by developing it and sharing it. RaptorJIT needs to preserve the qualities of LuaJIT that have drawn all of us into the community in the first place.

On the negative side, I do worry about the JIT being too opaque and mysterious to users. We need people to take the first step in understanding what the JIT does, and then the next step to identifying weaknesses, and the next step to fixing them, and the next step to reviewing and merging them, and so on. This is the pipeline that will lead to having an active maintainer community.

I am not sure exactly how to go about this but what I am doing for now is to share all my of thoughts here on Github, share experimental changes that may not pan out, and remove as much code and cleverness as possible without harming practical performance. I am confident that I am at least successfully demonstrating that I am not a super genius :) and I hope that myself and everybody following this repo is *gradually* developing the feeling that RaptorJIT is just another C program (and not a terribly big one at that.)

On the social side, I find that I don't actually relate to PUC Lua and the wider Lua community very much. On the one hand I am grateful that they created and documented a language that is simple, pleasant, and inoffensive. I would also like to make the RaptorJIT codebase as accessible as PUC Lua is. On the other hand I have never used PUC Lua myself and the software that I write absolutely depends on the JIT and FFI and so on. So I don't currently see divergence from the main line of PUC Lua development as a major problem, but I am open to feedback on that point.

**Comment by szbnwer** [comment]
hi there! :)

about _ENV i think there can be something like a local _ENV= <versioncheck> and _ENV or <new _ENV with metatable magic> - however ive never played with the env...

about hard to understand stuffs i think the best is to make shallow models for all the layers with the most basic informations to give a basic image with the "good to know details" from the perspective of a bird and with spoilers (hidden areas with toggle behaviour) with hardcore details and external readings; and pointers to deeper levels. the whole is in a tree graph with cross references for basic knowledge and deeper levels... the root is a general overview the leafs are mostly codeblocks... there can be labels for what level you are on - this way you can handle anything, all the info and codes are interconnected, any handlers can be used on any levels and whatever, maybe even a wiki can serve like this, but i think thats not the best for codes included and i could imagine a more powerful tool than that... im aiming these in my project (currently its still private and early alpha with slow development) however ive got a lotsa ideas around these if you are interested...


some off topic:
im interested in raptorjit but still not tried it. im making currently a tool that is for composing different bases for my project like experimental or stable and for setups for whatever purposes without messing with working bases, whenever ive done with that ill write the stuffs for using raptorjit as well, however im following the rj development as its interesting and ive learnt much from it even if i still havent make myself familiar with the codebase of lj nor rj, just read a bit from here and there from lj...

ive got some questions:
where can i go with offtopic, like email, irc or whatever?
why you have made studio with smalltalk (if im right...) instead of lua?
id suggest you to switch to a lua based toolkit (if nothing is against that), my favourite is tekui, however there are great competitors out there as well, but its better to use lua for making it more accessible for the lua community instead of using an another language, i think it gives more wind to your ship...
are studio and nix required to use, or only for introspecting?

and just came into my mind, ill leave it here before ill forget it: what about force the jit to blacklist or mark hot a codeblock if it can do so - if its a good idea then it could be used as well by lj

sorry for the off, i hope at least i could say anything useful :D
and all the bests for the ppl around! :)

**Comment by lukego** [comment]
> where can i go with offtopic, like email, irc or whatever?

Writing issues/comments here on Github is pretty good actually. This has the widest distribution, and too much discussion is better than too little. If we start to overflow then we can expand into other forums (like Slack #84 which is more like a water cooler).

> why you have made studio with smalltalk (if im right...) instead of lua?

Good question. Here is some context:

- Overall the diagnostic functions are split into two halves (#22). RaptorJIT _produces_ diagnostic data and other tools _analyze_ the data.
- RaptorJIT dumps data in a very raw format. This is to maximize the potential for post-hoc analysis (including developing new tools to use on old data.)
- There can be any number of analysis tools and they can be written in any programming language. Lua, perl, awk, R, Smalltalk, Javascript, etc, or any combination of the above. Compatibility layers could be built for existing tools like `jit.v` and `jit.dump` too (https://github.com/raptorjit/raptorjit/issues/22#issuecomment-287033967).
- My own experience of analyzing dumps makes me want a fancy tool like [Intel VTune](https://software.intel.com/en-us/intel-vtune-amplifier-xe). I want a unified interactive front-end for fluid navigation, I want easy access to advanced statistical tests to tackle non-determinism ("did this change _really_ improve the 10th percentile?"), and I want easy extensibility with Lua/Python/R/shell/etc too.
- Studio is the first analysis tool and it is based on this vision. Studio is not necessarily the only tool, or the best tool, or everybody's favorite tool, but it will at least be a damned practical tool :).
- RaptorJIT's diagnostic interface can evolve over time. Just for now it is raw C structs with DWARF metadata. This can be revised to help support more tools in the future.

Coming back to the question... There are several reasons that I chose Pharo+Nix for Studio instead of Lua:

- I use different languages for different problems. Lua is what I use for low-level system programming and I have a strong preference for Lua in this domain. However, I am happy to use other tools in other domains like GUIs, statistics, text processing, test suites, configuration management, etc.
- I am impressed with the [Glamorous Toolkit](http://gtoolkit.org/). This is an interactive object inspector, supporting plug-in domain-specific customized views, with a very active and helpful developer community, and with plenty of documentation. I would prefer to reuse this work, and participate in its existing community, rather than start from scratch.
- I hope that Studio will be used for many more projects besides RaptorJIT. So I don't want to assume that Studio developers will already be Lua hackers: they could equally well prefer C++, Go, Rust, Javascript, etc. I can't choose any technology that everybody will already know, so instead I pick ones that I know are worth learning :-).

> id suggest you to switch to a lua based toolkit [...] for making it more accessible for the lua community

This would make sense if Studio would be for RaptorJIT only. Then it could be worth building all the tools in Lua even if that requires more work up front, because it makes onboarding new developers easier over time. However, I want Studio to be used by many projects, and my strategy is to try and make it so great that enough people will be willing to learn Pharo+Nix to contribute. (Consider Emacs as a success story: Nobody knew Emacs Lisp before they started using Emacs, but people were motivated to make extensions and now look at all the _amazing_ stuff that has been developed there.)

So...

That is the background of RaptorJIT's diagnostic interface and the Studio tooling. Could well be that there is also demand for pure Lua tools and that this can be developed in parallel with Studio (and RaptorJIT can be evolved to support both.)

> are studio and nix required to use, or only for introspecting?

RaptorJIT does not depend on Studio or Nix. This is separate tooling that can be used to analyze RaptorJIT binary log files. So it's an arms-length model compared with `jit.v` and `jit.dump` that need to run inside the virtual machine while the application is running.

> what about force the jit to blacklist or mark hot a codeblock if it can do so - if its a good idea then it could be used as well by

Yeah. I don't know about these specific mechanisms but somehow users have to feel that they are in the driver's seat when it comes to JIT decisions. You cannot be at the mercy of guesswork heuristics to make your program run properly. See also https://github.com/LuaJIT/LuaJIT/issues/208.

**Comment by lukego** [comment]
Here is another related blog post: [Goodbye, Lua](https://realmensch.org/2016/05/28/goodbye-lua/). The author talks about a bunch of issues and this jumps out at me:

> LuaJIT will continue to be an awesome example of what a single brilliant developer can create, of course, but it means that the people who replace [Mike Pall] have huge shoes to fill. Odds are good that anyone who does step up (and I'm not following the status; people may have stepped up already, I don't know) will probably just maintain LuaJIT, fixing bugs, and at best may bring LuaJIT 2.1 to a final release state.

This is exactly the problem that RaptorJIT is intended to avoid. There is _so much more to do_ on LuaJIT and _so many more people_ who could contribute to the project. RaptorJIT is a project for these new people to do this new work :). We got 99 problems but stagnation ain't one...
